nestjs/src/gratitudes/gratitudes.service.ts

```diff
@@ -51,6 +52,61 @@ export class GratitudesService {
     });
   }
 
+  public async getHeroesRadar({ courseId, current: page = 1, pageSize = 20, notActivist }: HeroesRadarQueryDto) {
+    const countSubQuery = this.repository.createQueryBuilder('feedback');
+    const countQuery = this.dataSource.createQueryBuilder();
+
+    const heroesSubQuery = this.repository.createQueryBuilder('feedback');
+    const heroesQuery = this.dataSource.createQueryBuilder();
+
+    if (notActivist) {
+      [countSubQuery, heroesQuery].forEach(query => query.having(`not bool_or("badgeId" = 'RS_activist')`));
+    }
+
+    if (courseId) {
+      [heroesSubQuery, countSubQuery].forEach(query => query.where('feedback."courseId" = :courseId', { courseId }));
+    }
+
+    countSubQuery.select(`jsonb_agg(json_build_object('badgeId', "badgeId"))`).groupBy('"toUserId"');
+    countQuery
+      .select('COUNT(*) as count')
+      .from('(' + countSubQuery.getQuery() + ')', 'badges')
+      .setParameters(countSubQuery.getParameters());
+
+    heroesSubQuery
+      .select(['"feedback"."badgeId"', '"feedback"."toUserId"', 'COUNT(*) as "badgeCount"'])
+      .groupBy('"badgeId"')
+      .addGroupBy('"toUserId"');
+
+    heroesQuery
+      .select([
+        '"user"."githubId"',
+        '"user"."firstName"',
+        '"user"."lastName"',
+        'sum("badgeCount") as total',
+        `jsonb_agg(json_build_object('badgeId', "badgeId", 'badgeCount', "badgeCount")) as badges`,
+      ])
+      .from('(' + heroesSubQuery.getQuery() + ')', 'badges')
+      .leftJoin('user', 'user', 'badges."toUserId" = "user"."id"')
+      .groupBy('"githubId"')
+      .addGroupBy('"firstName"')
+      .addGroupBy('"lastName"')
+      .orderBy('total', 'DESC')
+      .limit(pageSize)
+      .offset((page - 1) * pageSize)
+      .setParameters(heroesSubQuery.getParameters());
+
+    const { count } = await countQuery.getRawOne();
+    const total = +count;
+    const items = await heroesQuery.getRawMany();
+    const totalPages = Math.ceil(total / pageSize);
+
+    return {
+      items,
+      meta: { itemCount: items.length, total, current: page, pageSize, totalPages },
+    };
```

Comment:
Try using what we use for pagination everywhere:
https://github.com/rolling-scopes/rsschool-app/blob/master/nestjs/src/core/paginate/index.ts#L3

Reply:
> Try using what we use for pagination everywhere: https://github.com/rolling-scopes/rsschool-app/blob/master/nestjs/src/core/paginate/index.ts#L3

It can't be used here, because it's raw query